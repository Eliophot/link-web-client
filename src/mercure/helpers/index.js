import { EventSourcePolyfill as EventSource } from 'event-source-polyfill';

export const bindToMercureTopic = (mercureInfo, realTimeUpdates, topic, onMessage, onTokenExpired) => () => {
  const { enabled } = realTimeUpdates;
  const { mercureHubUrl, token, loading, error } = mercureInfo;

  if (!enabled || loading || error) {
    return undefined;
  }

  const hubUrl = new URL(mercureHubUrl);

  hubUrl.searchParams.append('topic', topic);
  const es = new EventSource(hubUrl, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  es.onmessage = ({ data }) => onMessage(JSON.parse(data));
  es.onerror = ({ status }) => status === 401 && onTokenExpired();

  return () => es.close();
};
