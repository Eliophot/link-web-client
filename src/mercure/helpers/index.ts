import { EventSourcePolyfill as EventSource } from 'event-source-polyfill';
import { MercureInfo } from '../reducers/mercureInfo';

export const bindToMercureTopic = <T>(mercureInfo: MercureInfo, topic: string, onMessage: (message: T) => void, onTokenExpired: Function) => { // eslint-disable-line max-len
  const { mercureHubUrl, token, loading, error } = mercureInfo;

  if (loading || error || !mercureHubUrl) {
    return undefined;
  }

  const hubUrl = new URL(mercureHubUrl);

  hubUrl.searchParams.append('topic', topic);
  const es = new EventSource(hubUrl, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  es.onmessage = ({ data }: { data: string }) => onMessage(JSON.parse(data) as T);
  es.onerror = ({ status }: { status: number }) => status === 401 && onTokenExpired();

  return () => es.close();
};
