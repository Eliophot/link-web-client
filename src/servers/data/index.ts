export interface ServerData {
  name: string;
  url: string;
  apiKey: string;
}

export interface ServerWithId extends ServerData {
  id: string;
}

export interface ReachableServer extends ServerWithId {
  version: string;
  printableVersion: string;
}

export interface NonReachableServer extends ServerWithId {
  serverNotReachable: true;
}

export interface NotFoundServer {
  serverNotFound: true;
}

export type RegularServer = ReachableServer | NonReachableServer;

export type SelectedServer = RegularServer | NotFoundServer | null;

export type ServersMap = Record<string, ServerWithId>;

export const hasServerData = (server: ServerData | NotFoundServer | null): server is ServerData =>
  !!(server as ServerData)?.url && !!(server as ServerData)?.apiKey;

export const isReachableServer = (server: SelectedServer): server is ReachableServer =>
  !!server?.hasOwnProperty('printableVersion');

export const isServerWithId = (server: SelectedServer | ServerWithId): server is ServerWithId =>
  !!server?.hasOwnProperty('id');
