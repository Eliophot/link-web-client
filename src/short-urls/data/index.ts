import * as m from 'moment';
import { Nullable, OptionalString } from '../../utils/utils';

export interface EditShortUrlData {
  longUrl?: string;
  tags?: string[];
  title?: string;
  validSince?: m.Moment | string | null;
  validUntil?: m.Moment | string | null;
  maxVisits?: number | null;
  validateUrl?: boolean;
}

export interface ShortUrlData extends EditShortUrlData {
  longUrl: string;
  customSlug?: string;
  shortCodeLength?: number;
  domain?: string;
  findIfExists?: boolean;
}

export interface ShortUrl {
  shortCode: string;
  shortUrl: string;
  longUrl: string;
  dateCreated: string;
  visitsCount: number;
  meta: Required<Nullable<ShortUrlMeta>>;
  tags: string[];
  domain: string | null;
  title?: string | null;
}

export interface ShortUrlMeta {
  validSince?: string;
  validUntil?: string;
  maxVisits?: number;
}

export interface ShortUrlModalProps {
  shortUrl: ShortUrl;
  isOpen: boolean;
  toggle: () => void;
}

export interface ShortUrlIdentifier {
  shortCode: string;
  domain: OptionalString;
}
