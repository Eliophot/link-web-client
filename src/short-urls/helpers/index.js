import { isNil } from 'ramda';

export const shortUrlMatches = (shortUrl, shortCode, domain) => {
  if (isNil(domain)) {
    return shortUrl.shortCode === shortCode && !shortUrl.domain;
  }

  return shortUrl.shortCode === shortCode && shortUrl.domain === domain;
};
