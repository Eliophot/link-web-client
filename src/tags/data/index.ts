import type { ShlinkVisitsSummary } from '../../api/types';

export interface TagStats {
  shortUrlsCount: number;
  visitsCount: number;
  visitsSummary?: ShlinkVisitsSummary; // Optional only before Shlink 3.5.0
}

export interface TagModalProps {
  tag: string;
  isOpen: boolean;
  toggle: () => void;
}

export interface SimplifiedTag {
  tag: string;
  shortUrls: number;
  visits: number;
}
