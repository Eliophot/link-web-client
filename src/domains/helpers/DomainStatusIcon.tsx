import { FC, useRef } from 'react';
import { UncontrolledTooltip } from 'reactstrap';
import { ExternalLink } from 'react-external-link';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faBan as forbiddenIcon,
  faCheck as checkIcon,
  faCircleNotch as loadingStatusIcon,
} from '@fortawesome/free-solid-svg-icons';
import { DomainStatus } from '../data';

export const DomainStatusIcon: FC<{ status: DomainStatus }> = ({ status }) => {
  const ref = useRef<HTMLSpanElement>();

  if (status === 'validating') {
    return <FontAwesomeIcon fixedWidth icon={loadingStatusIcon} spin />;
  }

  return (
    <>
      <span
        ref={(el: HTMLSpanElement) => {
          ref.current = el;
        }}
      >
        {status === 'valid'
          ? <FontAwesomeIcon fixedWidth icon={checkIcon} className="text-muted" />
          : <FontAwesomeIcon fixedWidth icon={forbiddenIcon} className="text-danger" />}
      </span>
      <UncontrolledTooltip target={(() => ref.current) as any} placement="bottom" autohide={status === 'valid'}>
        {status === 'valid' ? 'Congratulations! This domain is properly configured.' : (
          <>
            Oops! There is some missing configuration, and short URLs shared with this domain will not work.
            <br />
            Follow <ExternalLink href="https://slnk.to/multi-domain-docs">the documentation</ExternalLink> in order to
            find out what is missing.
          </>
        )}
      </UncontrolledTooltip>
    </>
  );
};
