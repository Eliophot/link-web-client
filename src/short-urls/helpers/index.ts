import { isNil } from 'ramda';
import { ShortUrl, ShortUrlData } from '../data';
import { OptionalString } from '../../utils/utils';
import { DEFAULT_DOMAIN } from '../../visits/reducers/domainVisits';
import { ShortUrlCreationSettings } from '../../settings/reducers/settings';
import { isBefore, parseISO } from 'date-fns';
import { now } from '../../utils/helpers/date';

export const shortUrlMatches = (shortUrl: ShortUrl, shortCode: string, domain: OptionalString): boolean => {
  if (isNil(domain)) {
    return shortUrl.shortCode === shortCode && !shortUrl.domain;
  }

  return shortUrl.shortCode === shortCode && shortUrl.domain === domain;
};

export const domainMatches = (shortUrl: ShortUrl, domain: string): boolean => {
  if (!shortUrl.domain && domain === DEFAULT_DOMAIN) {
    return true;
  }

  return shortUrl.domain === domain;
};

export const shortUrlIsDisabled = (shortUrl: ShortUrl): boolean => (
  !!shortUrl.meta.maxVisits && shortUrl.visitsCount >= shortUrl.meta.maxVisits
) || (
  !!shortUrl.meta.validSince && isBefore(now(), parseISO(shortUrl.meta.validSince))
) || (
  !!shortUrl.meta.validUntil && isBefore(parseISO(shortUrl.meta.validUntil), now())
);

export const shortUrlDataFromShortUrl = (shortUrl?: ShortUrl, settings?: ShortUrlCreationSettings): ShortUrlData => {
  const validateUrl = settings?.validateUrls ?? false;

  if (!shortUrl) {
    return { longUrl: '', validateUrl };
  }

  return {
    longUrl: shortUrl.longUrl,
    tags: shortUrl.tags,
    title: shortUrl.title ?? undefined,
    domain: shortUrl.domain ?? undefined,
    validSince: shortUrl.meta.validSince ?? undefined,
    validUntil: shortUrl.meta.validUntil ?? undefined,
    maxVisits: shortUrl.meta.maxVisits ?? undefined,
    crawlable: shortUrl.crawlable,
    forwardQuery: shortUrl.forwardQuery,
    validateUrl,
  };
};

const MULTI_SEGMENT_SEPARATOR = '__';

export const urlEncodeShortCode = (shortCode: string): string => shortCode.replaceAll('/', MULTI_SEGMENT_SEPARATOR);

export const urlDecodeShortCode = (shortCode: string): string => shortCode.replaceAll(MULTI_SEGMENT_SEPARATOR, '/');
