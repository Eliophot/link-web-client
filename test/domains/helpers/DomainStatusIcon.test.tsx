import { shallow, ShallowWrapper } from 'enzyme';
import { UncontrolledTooltip } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes, faCheck, faCircleNotch } from '@fortawesome/free-solid-svg-icons';
import { DomainStatus } from '../../../src/domains/data';
import { DomainStatusIcon } from '../../../src/domains/helpers/DomainStatusIcon';

describe('<DomainStatusIcon />', () => {
  let wrapper: ShallowWrapper;
  const createWrapper = (status: DomainStatus) => {
    wrapper = shallow(<DomainStatusIcon status={status} />);

    return wrapper;
  };

  afterEach(() => wrapper?.unmount());

  it('renders loading icon when status is "validating"', () => {
    const wrapper = createWrapper('validating');
    const tooltip = wrapper.find(UncontrolledTooltip);
    const faIcon = wrapper.find(FontAwesomeIcon);

    expect(tooltip).toHaveLength(0);
    expect(faIcon).toHaveLength(1);
    expect(faIcon.prop('icon')).toEqual(faCircleNotch);
    expect(faIcon.prop('spin')).toEqual(true);
  });

  it.each([
    [
      'invalid' as DomainStatus,
      faTimes,
      'Oops! There is some missing configuration, and short URLs shared with this domain will not work.',
    ],
    [ 'valid' as DomainStatus, faCheck, 'Congratulations! This domain is properly configured.' ],
  ])('renders expected icon and tooltip when status is not validating', (status, expectedIcon, expectedText) => {
    const wrapper = createWrapper(status);
    const tooltip = wrapper.find(UncontrolledTooltip);
    const faIcon = wrapper.find(FontAwesomeIcon);
    const getTooltipText = (): string => {
      const children = tooltip.prop('children');

      if (typeof children === 'string') {
        return children;
      }

      return tooltip.find('span').html();
    };

    expect(tooltip).toHaveLength(1);
    expect(tooltip.prop('autohide')).toEqual(status === 'valid');
    expect(getTooltipText()).toContain(expectedText);
    expect(faIcon).toHaveLength(1);
    expect(faIcon.prop('icon')).toEqual(expectedIcon);
    expect(faIcon.prop('spin')).toEqual(false);
  });
});
