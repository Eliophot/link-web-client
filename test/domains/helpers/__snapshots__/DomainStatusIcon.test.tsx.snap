// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DomainStatusIcon /> renders expected icon and tooltip when status is not validating 1`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-circle-notch fa-spin fa-fw "
  data-icon="circle-notch"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M222.7 32.15C227.7 49.08 218.1 66.9 201.1 71.94C121.8 95.55 64 169.1 64 255.1C64 362 149.1 447.1 256 447.1C362 447.1 448 362 448 255.1C448 169.1 390.2 95.55 310.9 71.94C293.9 66.9 284.3 49.08 289.3 32.15C294.4 15.21 312.2 5.562 329.1 10.6C434.9 42.07 512 139.1 512 255.1C512 397.4 397.4 511.1 256 511.1C114.6 511.1 0 397.4 0 255.1C0 139.1 77.15 42.07 182.9 10.6C199.8 5.562 217.6 15.21 222.7 32.15V32.15z"
    fill="currentColor"
  />
</svg>
`;

exports[`<DomainStatusIcon /> renders expected icon and tooltip when status is not validating 2`] = `
<span>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-xmark fa-fw text-danger"
    data-icon="xmark"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 320 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M310.6 361.4c12.5 12.5 12.5 32.75 0 45.25C304.4 412.9 296.2 416 288 416s-16.38-3.125-22.62-9.375L160 301.3L54.63 406.6C48.38 412.9 40.19 416 32 416S15.63 412.9 9.375 406.6c-12.5-12.5-12.5-32.75 0-45.25l105.4-105.4L9.375 150.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L160 210.8l105.4-105.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-105.4 105.4L310.6 361.4z"
      fill="currentColor"
    />
  </svg>
</span>
`;

exports[`<DomainStatusIcon /> renders expected icon and tooltip when status is not validating 3`] = `
<span>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-check fa-fw text-muted"
    data-icon="check"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 448 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M438.6 105.4C451.1 117.9 451.1 138.1 438.6 150.6L182.6 406.6C170.1 419.1 149.9 419.1 137.4 406.6L9.372 278.6C-3.124 266.1-3.124 245.9 9.372 233.4C21.87 220.9 42.13 220.9 54.63 233.4L159.1 338.7L393.4 105.4C405.9 92.88 426.1 92.88 438.6 105.4H438.6z"
      fill="currentColor"
    />
  </svg>
</span>
`;

exports[`<DomainStatusIcon /> renders proper tooltip based on state 1`] = `
<div
  class="tooltip-inner"
  role="tooltip"
>
  <span>
    Oops! There is some missing configuration, and short URLs shared with this domain will not work.
    <br />
    Check the 
    <a
      href="https://slnk.to/multi-domain-docs"
      rel="noopener noreferrer"
      target="_blank"
    >
      documentation
    </a>
     in order to find out what is missing.
  </span>
</div>
`;

exports[`<DomainStatusIcon /> renders proper tooltip based on state 2`] = `
<div
  class="tooltip-inner"
  role="tooltip"
>
  Congratulations! This domain is properly configured.
</div>
`;
